libc-bottom-half/cloudlibc/src/libc/unistd/write.c

```diff
@@ -2,11 +2,131 @@
 //
 // SPDX-License-Identifier: BSD-2-Clause
 
+#ifdef __wasilibc_use_wasip2
+#include <wasi/wasip2.h>
+#include <wasi/descriptor_table.h>
+#include <common/errors.h>
+#include <time.h>
+#else
 #include <wasi/api.h>
+#endif
 #include <errno.h>
 #include <unistd.h>
 
 ssize_t write(int fildes, const void *buf, size_t nbyte) {
+#ifdef __wasilibc_use_wasip2
+  streams_borrow_output_stream_t output_stream;
+  filesystem_own_input_stream_t input_stream;
+  bool ok = false;
+  bool create_new_stream = false;
+  filesystem_error_code_t error_code;
+  descriptor_table_entry_t* entry = 0;
+
+  // Check for stdin/stdout/stderr
+  if (fildes == 1)
+    output_stream = streams_borrow_output_stream(stdout_get_stdout());
+  else if (fildes == 2)
+    output_stream = streams_borrow_output_stream(stderr_get_stderr());
+  else {
+    // Translate the file descriptor to an internal handle
+    descriptor_table_get_ref(fildes, &entry);
+    if (entry->tag == DESCRIPTOR_TABLE_ENTRY_FILE_HANDLE) {
+      create_new_stream = true;
+
+      // Get the output stream
+      filesystem_own_output_stream_t stream_owned;
+      ok = filesystem_method_descriptor_write_via_stream(entry->file.file_handle,
```

Comment:
Can we cache a stream per file_handle to be used by write, rather than open a new one each write?

Reply:
The code does that already -- this code is handling the case where the stream hasn't been opened yet. The `else` clause (`} else if (entry->tag == DESCRIPTOR_TABLE_ENTRY_FILE_STREAM) {` handles the case where there's an already-open stream.

Reply:
Thanks! I misunderstood that